src/Servers/Kestrel/Core/src/Internal/Http/Http1OutputProducer.cs

```diff
@@ -541,9 +541,13 @@ public void Reset()
         // These methods are for chunked http responses that use GetMemory/Advance
         private Memory<byte> GetChunkedMemory(int sizeHint)
         {
-            // The max size of a chunk will be the size of memory returned from the PipeWriter (today 4096)
-            // minus 5 for the max chunked prefix size and minus 2 for the chunked ending, leaving a total of
-            // 4089.
+            // Calling GetMemory on the PipeWriter can return a size larger than 4096
+            // if the sizeHint is greater.
+            // The max size we allow for chunking is 4096.
+            if (sizeHint > _memoryPool.MaxBufferSize)
```

Comment:
Hmm given the new behavior of GetMemory, should we use a heap buffer if it’s bigger than the pool size?

Reply:
That's the thing I'm contemplating. My list above states that either are valid options, just the second one requires more work.

Reply:
So the problem with this change is that the HTTP/1 response PipeWriter once again return less than sizeHint bytes from GetMemory even though the new contract is that PipeWriters always return at least sizeHint bytes? 

If so, I can see how proper fix could be a lot more involved than this PR as is. We probably still don't want to send up to 2GB chunks, so it could require creating multiple chunks (and the doing the requisite copying) during a single call to Advance(bytes).

That was the nice thing about sizeHint being a hint instead of a minimum: If the PipeWriter implementation knows it's going to have to chunk the data anyway, it can induce the code using the PipeWriter to do the chunking in a way that will likely be more efficient.

Reply:
To confirm before I try to prototype it, should we make it so the sizeHint can be greater than 4096?

Reply:
I would say so. Let's wait to hear what @davidfowl has to say though.

Reply:
Yes we have to respect the contract and make it a minimum. We should just change all of the calling code (like IBufferWriter\<byte\>.Write) to call GetMemory with a smaller size buffer (like 4K) to avoid this happening in really exceptional cases. If the caller request 2GB then just throw OOM (or let the array construction do that)

Reply:
Or just calling GetMemory without a sizeHint.

Reply:
>  If the caller request 2GB then just throw OOM (or let the array construction do that)

The bigger concern is when array construction *doesn't* OOM. There are likely at least some clients/proxies that limit received HTTP chunks to a max size of somewhere between 4KB and 2GB. That's why I'm concerned about making a change that allows for greater-than-4KB chunks for the first time.

Example: https://github.com/softwaremill/sttp/issues/36

Reply:
Then we can throw ourselves. It should be a setting though

Reply:
> It should be a setting though

👍

Reply:
> The bigger concern is when array construction doesn't OOM. There are likely at least some clients/proxies that limit received HTTP chunks to a max size of somewhere between 4KB and 2GB. That's why I'm concerned about making a change that allows for greater-than-4KB chunks for the first time.

I'm fairly certain this 4KB restriction is only there for the PipeWriter scenario. If someone called Response.WriteAsync in the past, I'm fairly sure it would write a single chunk for that write. Ex if someone does `ctx.Response.Body.WriteAsync(2gb string)`, we would write that chunk to the response still.

If we want to throw/limit chunk size, we should do it universally.

Reply:
🆙 📅 without throwing for OOM.